File: app/src/main/java/com/sandy/agorachatsandy/activites/AgoraMessageActivity.java
package com.sandy.agorachatsandy.activites;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.sandy.agorachatsandy.R;
import com.sandy.agorachatsandy.adapter.AgoraMessageAdapter;
import com.sandy.agorachatsandy.model.AgoraMessageBean;
import com.sandy.agorachatsandy.model.AgoraUser;
import com.sandy.agorachatsandy.model.MessageListBean;
import com.sandy.agorachatsandy.rtm.AGApplication;
import com.sandy.agorachatsandy.rtm.AgoraChatManager;
import com.sandy.agorachatsandy.utils.MessageUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.agora.rtm.ErrorInfo;
import io.agora.rtm.ResultCallback;
import io.agora.rtm.RtmChannel;
import io.agora.rtm.RtmChannelAttribute;
import io.agora.rtm.RtmChannelListener;
import io.agora.rtm.RtmChannelMember;
import io.agora.rtm.RtmClient;
import io.agora.rtm.RtmClientListener;
import io.agora.rtm.RtmMessage;
import io.agora.rtm.RtmStatusCode;

public class AgoraMessageActivity extends AppCompatActivity {

    private final String TAG = AgoraMessageActivity.class.getSimpleName();

    private TextView mTitleTextView;
    private EditText mMsgEditText;
    private RecyclerView mRecyclerView;
    private List<AgoraMessageBean> mMessageBeanList = new ArrayList<>();
    private AgoraMessageAdapter mMessageAdapter;

    private boolean mIsPeerToPeerMode = true;
    private AgoraUser user;
    private String mPeerId = "";
    private String mChannelName = "";
    private int mChannelMemberCount = 1;
    private String targetName = "";

    private AgoraChatManager mChatManager;
    private RtmClient mRtmClient;
    private RtmClientListener mClientListener;
    private RtmChannel mRtmChannel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.agora_activity_message);
        init();
    }

    private void init() {
        mChatManager = AGApplication.the().getChatManager();
        mRtmClient = mChatManager.getRtmClient();
        mClientListener = new MyRtmClientListener();
        mChatManager.registerListener(mClientListener);

        getExtras();

        mTitleTextView = findViewById(R.id.message_title);

        if (mIsPeerToPeerMode) {
            mPeerId = targetName;
            mTitleTextView.setText(mPeerId);

            // load history chat records
            MessageListBean messageListBean = MessageUtil.getExistMessageListBean(mPeerId);
            if (messageListBean != null) {
                mMessageBeanList.addAll(messageListBean.getMessageBeanList());
            }

            // load offline messages since last chat with this peer.
            // Then clear cached offline messages from message pool
            // since they are already consumed.
            MessageListBean offlineMessageBean = new MessageListBean(mPeerId, mChatManager);
            mMessageBeanList.addAll(offlineMessageBean.getMessageBeanList());
            mChatManager.removeAllOfflineMessages(mPeerId);
        } else {
            mChannelName = targetName;
            mChannelMemberCount = 1;
            mTitleTextView.setText(mChannelName + "(" + mChannelMemberCount + ")");
            createAndJoinChannel();
        }

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(RecyclerView.VERTICAL);
        mMessageAdapter = new AgoraMessageAdapter(this, mMessageBeanList);
        mRecyclerView = findViewById(R.id.message_list);
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.setAdapter(mMessageAdapter);

        mMsgEditText = findViewById(R.id.message_edittiext);
    }

    private void getExtras() {
        Intent intent = getIntent();
        mIsPeerToPeerMode = intent.getBooleanExtra(MessageUtil.INTENT_EXTRA_IS_PEER_MODE, true);
        user = intent.getParcelableExtra(MessageUtil.INTENT_EXTRA_USER_ID);
        targetName = intent.getStringExtra(MessageUtil.INTENT_EXTRA_TARGET_NAME);
    }

    /**
     * API CALL: create and join channel
     */
    private void createAndJoinChannel() {
        // step 1: create a channel instance
        mRtmChannel = mRtmClient.createChannel(mChannelName, new MyChannelListener());
        if (mRtmChannel == null) {
            showToast(getString(R.string.join_channel_failed));
            finish();
            return;
        }

        Log.e("channel", mRtmChannel + "");

        // step 2: join the channel
        mRtmChannel.join(new ResultCallback<Void>() {
            @Override
            public void onSuccess(Void responseInfo) {
                Log.i(TAG, "join channel success");
                getChannelMemberList();
            }

            @Override
            public void onFailure(ErrorInfo errorInfo) {
                Log.e(TAG, "join channel failed");
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        showToast(getString(R.string.join_channel_failed));
                        finish();
                    }
                });
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mIsPeerToPeerMode) {
            MessageUtil.addMessageListBeanList(new MessageListBean(mPeerId, mMessageBeanList));
        } else {
            leaveChannel();
        }
        mChatManager.unregisterListener(mClientListener);
    }

    /**
     * API CALL: leave and release channel
     */
    private void leaveChannel() {
        if (mRtmChannel != null) {
            mRtmChannel.leave(new ResultCallback<Void>() {
                @Override
                public void onSuccess(Void aVoid) {

                }

                @Override
                public void onFailure(ErrorInfo errorInfo) {
                    showToast("leve channel failed");
                }
            });
            mRtmChannel = null;
        }
    }

    /**
     * API CALL: get channel member list
     */
    private void getChannelMemberList() {
        mRtmChannel.getMembers(new ResultCallback<List<RtmChannelMember>>() {
            @Override
            public void onSuccess(final List<RtmChannelMember> responseInfo) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        mChannelMemberCount = responseInfo.size();
                        refreshChannelTitle();
                    }
                });
            }

            @Override
            public void onFailure(ErrorInfo errorInfo) {
                Log.e(TAG, "failed to get channel members, err: " + errorInfo.getErrorCode());
            }
        });
    }

    public void onClickSend(View v) {
        String msg = mMsgEditText.getText().toString();
        if (!msg.equals("")) {
            AgoraMessageBean messageBean = new AgoraMessageBean(user.getFireDisplayName(), msg, true);
            mMessageBeanList.add(messageBean);
            mMessageAdapter.notifyItemRangeChanged(mMessageBeanList.size(), 1);
            mRecyclerView.scrollToPosition(mMessageBeanList.size() - 1);
            if (mIsPeerToPeerMode) {
                sendPeerMessage(msg);
            } else {
                sendChannelMessage(msg);
            }
        }
        mMsgEditText.setText("");
    }

    private void showToast(final String text) {
        Toast.makeText(AgoraMessageActivity.this, text, Toast.LENGTH_SHORT).show();
    }

    public void onClickFinish(View v) {
        finish();
    }

    /**
     * API CALLBACK: rtm channel event listener
     */
    class MyChannelListener implements RtmChannelListener {
        @Override
        public void onMemberCountUpdated(int i) {

        }

        @Override
        public void onAttributesUpdated(List<RtmChannelAttribute> list) {

        }

        @Override
        public void onMessageReceived(final RtmMessage message, final RtmChannelMember fromMember) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    String account = fromMember.getUserId();
                    String msg = message.getText();
                    Log.i(TAG, "onMessageReceived account = " + account + " msg = " + msg);
                    AgoraMessageBean messageBean = new AgoraMessageBean(account, msg, false);
                    messageBean.setBackground(getMessageColor(account));
                    mMessageBeanList.add(messageBean);
                    mMessageAdapter.notifyItemRangeChanged(mMessageBeanList.size(), 1);
                    mRecyclerView.scrollToPosition(mMessageBeanList.size() - 1);
                }
            });
        }

        @Override
        public void onMemberJoined(RtmChannelMember member) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mChannelMemberCount++;
                    refreshChannelTitle();
                }
            });
        }

        @Override
        public void onMemberLeft(RtmChannelMember member) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mChannelMemberCount--;
                    refreshChannelTitle();
                }
            });
        }
    }

    /**
     * API CALL: send message to a channel
     */
    private void sendChannelMessage(String content) {
        // step 1: create a message
        RtmMessage message = mRtmClient.createMessage();
        message.setText(content);

        Log.e("channel", mRtmChannel + "");

        // step 2: send message to channel
        mRtmChannel.sendMessage(message, new ResultCallback<Void>() {
            @Override
            public void onSuccess(Void aVoid) {

            }

            @Override
            public void onFailure(ErrorInfo errorInfo) {
                // refer to RtmStatusCode.ChannelMessageState for the message state
                final int errorCode = errorInfo.getErrorCode();
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        switch (errorCode) {
                            case RtmStatusCode.ChannelMessageError.CHANNEL_MESSAGE_ERR_TIMEOUT:
                            case RtmStatusCode.ChannelMessageError.CHANNEL_MESSAGE_ERR_FAILURE:
                                showToast(getString(R.string.send_msg_failed));
                                break;
                        }
                    }
                });
            }
        });
    }

    /**
     * API CALL: send message to peer
     */
    private void sendPeerMessage(String content) {
        // step 1: create a message
        RtmMessage message = mRtmClient.createMessage();
        message.setText(content);

        // step 2: send message to peer
        mRtmClient.sendMessageToPeer(mPeerId, message, mChatManager.getSendMessageOptions(), new ResultCallback<Void>() {
            @Override
            public void onSuccess(Void aVoid) {

            }

            @Override
            public void onFailure(ErrorInfo errorInfo) {
                // refer to RtmStatusCode.PeerMessageState for the message state
                final int errorCode = errorInfo.getErrorCode();
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        switch (errorCode) {
                            case RtmStatusCode.PeerMessageError.PEER_MESSAGE_ERR_TIMEOUT:
                            case RtmStatusCode.PeerMessageError.PEER_MESSAGE_ERR_FAILURE:
                                showToast(getString(R.string.send_msg_failed));
                                break;
                            case RtmStatusCode.PeerMessageError.PEER_MESSAGE_ERR_PEER_UNREACHABLE:
                                showToast(getString(R.string.peer_offline));
                                break;
                            case RtmStatusCode.PeerMessageError.PEER_MESSAGE_ERR_CACHED_BY_SERVER:
                                showToast(getString(R.string.message_cached));
                                break;
                        }
                    }
                });
            }
        });
    }

    private int getMessageColor(String account) {
        for (int i = 0; i < mMessageBeanList.size(); i++) {
            if (account.equals(mMessageBeanList.get(i).getAccount())) {
                return mMessageBeanList.get(i).getBackground();
            }
        }
        return MessageUtil.COLOR_ARRAY[MessageUtil.RANDOM.nextInt(MessageUtil.COLOR_ARRAY.length)];
    }

    private void refreshChannelTitle() {
        String titleFormat = getString(R.string.channel_title);
        String title = String.format(titleFormat, mChannelName, mChannelMemberCount);
        mTitleTextView.setText(title);
    }

    class MyRtmClientListener implements RtmClientListener {
        @Override
        public void onPeersOnlineStatusChanged(Map<String, Integer> map) {

        }

        @Override
        public void onConnectionStateChanged(final int state, int reason) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    switch (state) {
                        case RtmStatusCode.ConnectionState.CONNECTION_STATE_RECONNECTING:
                            showToast(getString(R.string.reconnecting));
                            break;
                        case RtmStatusCode.ConnectionState.CONNECTION_STATE_ABORTED:
                            showToast(getString(R.string.account_offline));
                            setResult(MessageUtil.ACTIVITY_RESULT_CONN_ABORTED);
                            finish();
                            break;
                    }
                }
            });
        }

        @Override
        public void onMessageReceived(final RtmMessage message, final String peerId) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    String content = message.getText();
                    if (peerId.equals(mPeerId)) {
                        AgoraMessageBean messageBean = new AgoraMessageBean(peerId, content, false);
                        messageBean.setBackground(getMessageColor(peerId));
                        mMessageBeanList.add(messageBean);
                        mMessageAdapter.notifyItemRangeChanged(mMessageBeanList.size(), 1);
                        mRecyclerView.scrollToPosition(mMessageBeanList.size() - 1);
                    } else {
                        MessageUtil.addMessageBean(peerId, content);
                    }
                }
            });
        }

        @Override
        public void onTokenExpired() {

        }
    }
}

File: app/src/main/java/com/sandy/agorachatsandy/activites/AgoraMessageActivity.kt
package com.sandy.agorachatsandy.activites

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.sandy.agorachatsandy.R
import com.sandy.agorachatsandy.activites.AgoraMessageActivity
import com.sandy.agorachatsandy.adapter.AgoraMessageAdapter
import com.sandy.agorachatsandy.model.AgoraMessageBean
import com.sandy.agorachatsandy.model.AgoraUser
import com.sandy.agorachatsandy.rtm.AGApplication
import com.sandy.agorachatsandy.rtm.AgoraChatManager
import com.sandy.agorachatsandy.utils.MessageUtil
import io.agora.rtm.*
import kotlinx.android.synthetic.main.agora_activity_message.*
import java.util.*

class AgoraMessageActivity : AppCompatActivity() {
    private val TAG = AgoraMessageActivity::class.java.simpleName
    private val mMessageBeanList: MutableList<AgoraMessageBean> = ArrayList()
    private var mMessageAdapter: AgoraMessageAdapter? = null
    private var user: AgoraUser? = null
    private var mChannelName = ""
    private var mChannelMemberCount = 1
    private var targetName = ""
    private var mChatManager: AgoraChatManager? = null

    private val mRtmClient: RtmClient by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
        val mChatManager = AGApplication.the().chatManager
        mChatManager.rtmClient
    }
    private var mClientListener: RtmClientListener? = null
    private var mRtmChannel: RtmChannel? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.agora_activity_message)
        val ab = supportActionBar
        ab?.hide()
        init()
    }

    private fun init() {
        user = intent.getParcelableExtra(MessageUtil.INTENT_EXTRA_USER_ID)
        targetName = intent.getStringExtra(MessageUtil.INTENT_EXTRA_TARGET_NAME)
        mChatManager = AGApplication.the().chatManager
        mClientListener = MyRtmClientListener()
        mChatManager?.registerListener(mClientListener)

        mChannelName = targetName
        mChannelMemberCount = 1
        message_title?.text = "$mChannelName($mChannelMemberCount)"
        createAndJoinChannel()
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = RecyclerView.VERTICAL
        mMessageAdapter = AgoraMessageAdapter(this, mMessageBeanList)
        message_list.layoutManager = layoutManager
        message_list.adapter = mMessageAdapter
    }

    /**
     * API CALL: create and join channel
     */
    private fun createAndJoinChannel() {
        // step 1: create a channel instance
        mRtmChannel = mRtmClient.createChannel(mChannelName, MyChannelListener())
        if (mRtmChannel == null) {
            showToast(getString(R.string.join_channel_failed))
            finish()
            return
        }
        Log.e("channel", mRtmChannel.toString() + "")

        // step 2: join the channel
        mRtmChannel!!.join(object : ResultCallback<Void?> {
            override fun onSuccess(responseInfo: Void?) {
                Log.i(TAG, "join channel success")
                channelMemberList
            }

            override fun onFailure(errorInfo: ErrorInfo) {
                Log.e(TAG, "join channel failed")
                runOnUiThread {
                    showToast(getString(R.string.join_channel_failed))
                    finish()
                }
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        leaveChannel()
        mChatManager!!.unregisterListener(mClientListener)
    }

    /**
     * API CALL: leave and release channel
     */
    private fun leaveChannel() {
        if (mRtmChannel != null) {
            mRtmChannel!!.leave(object : ResultCallback<Void?> {
                override fun onSuccess(aVoid: Void?) {}
                override fun onFailure(errorInfo: ErrorInfo) {
                    showToast("leve channel failed")
                }
            })
            mRtmChannel!!.release()
            mRtmChannel = null
        }
    }

    /**
     * API CALL: get channel member list
     */
    private val channelMemberList: Unit
        private get() {
            mRtmChannel!!.getMembers(object : ResultCallback<List<RtmChannelMember?>> {
                override fun onSuccess(responseInfo: List<RtmChannelMember?>) {
                    runOnUiThread {
                        mChannelMemberCount = responseInfo.size
                        refreshChannelTitle()
                    }
                }

                override fun onFailure(errorInfo: ErrorInfo) {
                    Log.e(TAG, "failed to get channel members, err: " + errorInfo.errorCode)
                }
            })
        }

    fun onClickSend(v: View?) {
        val msg = message_edittiext!!.text.toString()
        if (msg != "") {
            val messageBean = AgoraMessageBean(user!!.displayName, msg, true)
            mMessageBeanList.add(messageBean)
            mMessageAdapter!!.notifyItemRangeChanged(mMessageBeanList.size, 1)
            message_list!!.scrollToPosition(mMessageBeanList.size - 1)
            sendChannelMessage(msg)
        }
        message_edittiext!!.setText("")
    }

    private fun showToast(text: String) {
        Toast.makeText(this@AgoraMessageActivity, text, Toast.LENGTH_SHORT).show()
    }

    fun onClickFinish(v: View?) {
        finish()
    }

    /**
     * API CALLBACK: rtm channel event listener
     */
    internal inner class MyChannelListener : RtmChannelListener {
        override fun onMemberCountUpdated(i: Int) {}
        override fun onAttributesUpdated(list: List<RtmChannelAttribute>) {}
        override fun onMessageReceived(message: RtmMessage, fromMember: RtmChannelMember) {
            runOnUiThread {
                val account = fromMember.userId
                val msg = message.text
                Log.i(TAG, "onMessageReceived account = $account msg = $msg")
                val messageBean = AgoraMessageBean(account, msg, false)
                messageBean.background = getMessageColor(account)
                mMessageBeanList.add(messageBean)
                mMessageAdapter!!.notifyItemRangeChanged(mMessageBeanList.size, 1)
                message_list!!.scrollToPosition(mMessageBeanList.size - 1)
            }
        }

        override fun onMemberJoined(member: RtmChannelMember) {
            runOnUiThread {
                mChannelMemberCount++
                refreshChannelTitle()
            }
        }

        override fun onMemberLeft(member: RtmChannelMember) {
            runOnUiThread {
                mChannelMemberCount--
                refreshChannelTitle()
            }
        }
    }

    /**
     * API CALL: send message to a channel
     */
    private fun sendChannelMessage(content: String) {
        // step 1: create a message
        val message = mRtmClient.createMessage()
        message.text = content
        Log.e("channel", mRtmChannel.toString() + "")

        // step 2: send message to channel
        mRtmChannel!!.sendMessage(message, object : ResultCallback<Void?> {
            override fun onSuccess(aVoid: Void?) {}
            override fun onFailure(errorInfo: ErrorInfo) {
                // refer to RtmStatusCode.ChannelMessageState for the message state
                val errorCode = errorInfo.errorCode
                runOnUiThread {
                    when (errorCode) {
                        RtmStatusCode.ChannelMessageError.CHANNEL_MESSAGE_ERR_TIMEOUT, RtmStatusCode.ChannelMessageError.CHANNEL_MESSAGE_ERR_FAILURE -> showToast(getString(R.string.send_msg_failed))
                    }
                }
            }
        })
    }

    private fun getMessageColor(account: String): Int {
        for (i in mMessageBeanList.indices) {
            if (account == mMessageBeanList[i].account) {
                return mMessageBeanList[i].background
            }
        }
        return MessageUtil.COLOR_ARRAY[MessageUtil.RANDOM.nextInt(MessageUtil.COLOR_ARRAY.size)]
    }

    private fun refreshChannelTitle() {
        val titleFormat = getString(R.string.channel_title)
        val title = String.format(titleFormat, mChannelName, mChannelMemberCount)
        message_title!!.text = title
    }

    internal inner class MyRtmClientListener : RtmClientListener {
        override fun onPeersOnlineStatusChanged(map: Map<String, Int>) {}
        override fun onConnectionStateChanged(state: Int, reason: Int) {
            runOnUiThread {
                when (state) {
                    RtmStatusCode.ConnectionState.CONNECTION_STATE_RECONNECTING -> showToast(getString(R.string.reconnecting))
                    RtmStatusCode.ConnectionState.CONNECTION_STATE_ABORTED -> {
                        showToast(getString(R.string.account_offline))
                        setResult(MessageUtil.ACTIVITY_RESULT_CONN_ABORTED)
                        finish()
                    }
                }
            }
        }

        override fun onMessageReceived(message: RtmMessage, peerId: String) {
            runOnUiThread {
                val content = message.text
                if (peerId == "") {
                    val messageBean = AgoraMessageBean(peerId, content, false)
                    messageBean.background = getMessageColor(peerId)
                    mMessageBeanList.add(messageBean)
                    mMessageAdapter!!.notifyItemRangeChanged(mMessageBeanList.size, 1)
                    message_list!!.scrollToPosition(mMessageBeanList.size - 1)
                } else {
                    MessageUtil.addMessageBean(peerId, content)
                }
            }
        }

        override fun onTokenExpired() {}
    }
}